sam0_common: Add USB peripheral driver

bergzand opened the pull request:
### Contribution description

This PR adds a USB peripheral driver for the sam0 device class. The driver is much WIP as some details still have to be thought out.

### Testing procedure

Nothing yet, a follow-up is required to actually initialize the usb device.

### Issues/PRs references

depends on #9830

![image](https://bergzand.net/misc/usbus.dot.svg)

dylad commented on cpu/sam0_common/include/sam_usb.h:
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2018 Freie Universität Berlin
+ * Copyright (C) 2018 Inria
```
You should probably update these copyrights.

bergzand replied:
Fixed, thanks

dylad commented on cpu/sam0_common/include/sam_usb.h:
```diff
@@ -0,0 +1,64 @@
+ * @{
+ *
+ * @file
+ * @brief       [RFC 4122](https://tools.ietf.org/html/rfc4122) UUID functions
```
Please also update Doxygen

bergzand replied:
Fixed

dylad commented on cpu/sam0_common/include/sam_usb.h:
```diff
@@ -0,0 +1,64 @@
+#define SAM_USB_NUM_EP      USBDEV_NUM_ENDPOINTS
+
+typedef struct {
+    usbdev_t usbdev;
```
missing description

bergzand replied:
On the TODO

bergzand replied:
fixed

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,679 @@
+    return 2 * num + (dir == USB_EP_DIR_OUT ? 0 : 1);
+}
+
+static inline unsigned _get_ep_num2(usbdev_ep_t *ep)
```
Why is there two  get_ep_num functions ?

bergzand replied:
Probably just bad naming, but will check.

bergzand replied:
Refactored these a bit to make more sense

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,679 @@
+            val = 0x7;
+            break;
+        default:
+            return;
```
What about a single line to directly compute this value with a nice comment ? :)

bergzand replied:
Should be possible, but have to think about it a bit, the last case (1023) makes it a bit annoying.

bergzand replied:
I've been thinking a bit about this. I'd rather keep it this way as signalling error back in case of invalid size is easier to do with the switch case statement

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,679 @@
+    int res = -ENOTSUP;
+    assert(ep != NULL);
+    switch (opt) {
+        case USBOPT_EP_ENABLE:
```
add fall-through comment

dylad commented on boards/samr21-xpro/Makefile.dep:
```diff
@@ -5,3 +5,7 @@ endif
 endif
+
+ifneq (,$(filter usbdev,$(USEMODULE)))
+  USEMODULE += sam_usb
```
Why do you enable this feature for SAMR21 board only ?
You can put it directly in `cpu/sam0_common/Makefile.dep` instead.

bergzand replied:
> Why do you enable this feature for SAMR21 board only ?

My idea here was to only enable the the feature for boards that expose the USB port.

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+    /* Reset peripheral */
+    USB->DEVICE.CTRLA.reg |= USB_CTRLA_SWRST;
+    while (usb_swrst_syncing()) {}
+    while (USB->DEVICE.CTRLA.bit.SWRST) {}
```
You can remove this line. The datasheet states that both bit will be clearer once reset is done, so the line above should be enough.

bergzand replied:
Duplicate something something...

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+        USB_PADCAL_TRIM((*(uint32_t *)USB_FUSES_TRIM_ADDR >>
+                         USB_FUSES_TRIM_Pos));
+
+    USB->DEVICE.CTRLB.bit.SPDCONF = 0x0;
```
Replaces 0 by USB_DEVICE_CTRLB_SPDCONF_FS

bergzand replied:
Ack

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+            *(usb_speed_t *)value = USB_SPEED_FULL;
+            res = sizeof(usb_speed_t);
+            break;
+        default:
```
Maybe add a DEBUG info here ?

bergzand replied:
Done

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
```
Same here.

bergzand replied:
Also done

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+            ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_STALLRQ1;
+        }
+        else {
+            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_STALLRQ1;
```
```suggestion
            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSCLR_STALLRQ1;
```

bergzand replied:
fixed :blush:

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+            ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_STALLRQ0;
+        }
+        else {
+            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_STALLRQ0;
```
```suggestion
            ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSCLR_STALLRQ0;
```

bergzand replied:
fixed :blush:

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
```
DEBUG line ?

bergzand replied:
Added

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+    UsbDeviceEndpoint *ep_reg = &USB->DEVICE.DeviceEndpoint[ep->num];
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_BK1RDY;
```
```suggestion
        ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSCLR_BK1RDY;
```

bergzand replied:
fixed :blush:

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+        ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSSET_BK1RDY;
+    }
+    else {
+        ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSCLR_BK0RDY;
```
```suggestion
        ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_BK0RDY;
```

bergzand replied:
fixed :blush:

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+
+    /* Enable USB device */
+    USB->DEVICE.DESCADD.reg = (uint32_t)banks;
+    USB->DEVICE.CTRLA.reg |= USB_CTRLA_ENABLE | USB_CTRLA_RUNSTDBY;
```
I think you can remove `USB_CTRLA_RUNSTDBY` for now or make it user defined in `periph_conf.h`

bergzand replied:
Removed for now

dylad commented on drivers/include/usb/usbdev.h:
```diff
@@ -0,0 +1,337 @@
+struct usbdev {
+    const struct usbdev_driver *driver;     /**< usbdev driver struct   */
+    usbdev_event_cb_t cb;                   /**< Event callback supplied by
+                                              *  upper layer */
```
What about adding `UsbDevice` struct from atmel vendor files ? In the current state, this driver works fine with one USB device instance but maybe one day we will have several USB within a MCU.
This way we can replace all USB->DEVICE from the driver by a more generic call.

bergzand replied:
Agreed

bergzand replied:
I assume you want to add that to the `sam0_common_usb_t` and not to the generic `usbdev` struct.

I think your point is valid, but I have to check if the current API allows this. If it doesn't, I would assume this to be a shortcomming of the API.

dylad replied:
> I assume you want to add that to the `sam0_common_usb_t` and not to the generic `usbdev` struct.

Yes

bergzand replied:
Done, I'm not fully content with the current implementation for retrieving the context in the isr, but this can be reworked as soon as we actually have a device with multiple usb peripherals

dylad commented on cpu/sam0_common/usb/usb.c:
```diff
@@ -0,0 +1,671 @@
+{
+    (void)ep;
+    (void)value;
+    (void)max_len;
```
You can remove 
```
    (void)ep;
    (void)value;
```

bergzand replied:
Should be fixed now

bergzand commented:
Rebased and reworked to the current state of #9830 

Also fixed a number of issues/suggestions as proposed by @dylad

bergzand commented:
Removed WIP status btw, ready for review (I hope)

bergzand commented:
rebased

dylad commented:
@bergzand i'll give it another round this weekend. Could you rebase please ?

bergzand commented:
Reworked and rebased!

dylad commented on cpu/sam0_common/include/sam_usb.h:
```diff
@@ -0,0 +1,74 @@
+/**
+ * USB endpoint buffer space
+ *
+ * @todo global configurable?
```
still todo ?

bergzand replied:
Nope, actually is globally configurable now (for a while)

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+        case 512:
+            val = 0x6;
+            break;
+        case 1023:
```
@bergzand I think I've read somewhere that some endpoint size can be 1023 or 1024 in specific case. Am I wrong ?

bergzand replied:
1023 bytes is the max size for USB 2.0 full speed isochronous transfers, USB 2.0 high speed is allowed to have transfers of 1024 bytes with isochronous endpoint types.

dylad replied:
Alright then we're all good here :)

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_USB;
```
You missed `MCLK->APBBMASK` for SAML21 family

bergzand replied:
Thanks!

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
```
`#elif defined(CPU_FAM_SAML21) || defined(CPU_FAM_SAMR30)`

bergzand replied:
Thanks!

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+    gpio_init(GPIO_PIN(PA, 24), GPIO_IN);
+    gpio_init(GPIO_PIN(PA, 25), GPIO_IN);
+    gpio_init_mux(GPIO_PIN(PA, 24), GPIO_MUX_G);
+    gpio_init_mux(GPIO_PIN(PA, 25), GPIO_MUX_G);
```
You should reused your config struct defined in periph_conf.h here.

bergzand replied:
So much effort spent on the config struct, only to forget to use it in the implementation…

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+    usbdev->config->device->CTRLB.bit.SPDCONF = USB_DEVICE_CTRLB_SPDCONF_FS;
+    _enable_irq(usbdev);
+
+    gpio_init_int(GPIO_PIN(PA, 7), GPIO_IN_PD, GPIO_BOTH,
```
same comment as above

bergzand replied:
Fixed

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,739 @@
+    sam0_common_usb_t *usbdev = (sam0_common_usb_t*)dev;
+    if (usbdev->connect_change) {
+        usbdev->connect_change = false;
+        if (gpio_read(GPIO_PIN(PA, 7))) {
```
same comment as above

bergzand replied:
Also fixed

bergzand commented:
@dylad fixed your remarks, also uncrustified the file (but should have done that in a separate fixup commit :'( )

dylad commented:
> also uncrustified the file (but should have done that in a separate fixup commit :'( )

I don't mind, I'll test it on hardware asap so we can move forward.

bergzand commented:
> I don't mind, I'll test it on hardware asap so we can move forward.

Thanks, appreciating your reviews!

bergzand commented:
I messed up the GPIO things :(

bergzand commented:
> I messed up the GPIO things :(

fixed

bergzand commented:
@dylad I've reworked the driver with the ideas as discussed out-of-band today. Sorry for the force-push, messed up my commit order a bit.

- usbdev is extended with suspend and resume event types. I can split these out into a separate PR if you like
- the sam0_common usbdev driver makes use of these. Low power modes are unblocked when the device is suspended. Detaching the device from a host is also registered as a "suspend" here.
- I've added config for the arduino-zero board.

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,768 @@
+    return res;
+}
+
+static int _ep_unready(usbdev_ep_t *ep)
```
nitpick: why `_ep_unready` and `_usbdev_ep_ready` ?

bergzand replied:
Arbitrary, nothing specific, position of the moon etc.

Will fix to `_ep_ready`

bergzand replied:
Never mind the above, I remember now, all the `static _usbdev_$SOMETHING` calls are used in the instantiated `usbdev_driver_t` struct

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,768 @@
+static void _usbdev_ep_esr(usbdev_ep_t *ep)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+    signed event = -1;
```
int32 or int16 ? I'm sure this one will be troublesome outside of ARM.

bergzand replied:
But this driver is closely tied to an ARM based core…

I can change it to `int32_t` if you like

bergzand replied:
But same as [this comment](https://github.com/RIOT-OS/RIOT/pull/11213#discussion_r267553921), there is not really a need for a specific variable size here, so I'd rather give the compiler more freedom for possible optimizations.

bergzand commented on boards/samr21-xpro/include/periph_conf.h:
```diff
@@ -274,6 +274,21 @@ static const adc_conf_chan_t adc_channels[] = {
+        .dm     = GPIO_PIN(PA, 24),
+        .dm_mux = GPIO_MUX_G,
+        .dp     = GPIO_PIN(PA, 25),
+        .dp_mux = GPIO_MUX_G,
```
Ack, will fix

dylad commented on drivers/include/periph/usbdev.h:
```diff
@@ -158,6 +158,16 @@ typedef enum {
+    /**
+     * @brief USB suspend condition no longer active
+     */
+    USBDEV_EVENT_RESUME,
```
dump question here:
Is this suppose to be random or alphabetical order ?

bergzand replied:
Good question. The only "order" I had in mind was something that appeared logical to me. Which at the moment is to group the similar events together as much as possible.

dylad replied:
Alright, I don't think the coding guideline force us to change something here. Let's keep it this way.

bergzand commented:
rebased to current master

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,768 @@
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21) || defined(CPU_FAM_SAMR30)
```
I would change this line to
```suggestion
#elif defined(CPU_SAML21)
```
This will prevent us to add new sub-family like the upcoming SAMR34

bergzand replied:
Huh? Are you sure? I changed this to the current situation based on [your comment](https://github.com/RIOT-OS/RIOT/pull/10915#discussion_r266626774). I don't mind changing it back to what you are suggesting, just want to make sure that we get this right.

dylad replied:
You can go ahead, this is the way to go.

bergzand replied:
See 4b9b90bfcb12e888df4c01eda40323f0ea8a539f

dylad commented on cpu/sam0_common/periph/usbdev.c:
```diff
@@ -0,0 +1,768 @@
+        case USB_EP_TYPE_INTERRUPT:
+            type = 0x04;
+            break;
+        case USB_EP_TYPE_NONE:
```
add default statement at the same time here ?

bergzand replied:
How about I add an
```
default:
  assert(false);
```
here. Anything else than the types already handled in the switch is an error.

bergzand replied:
Never mind, I can just change the `case USB_EP_TYPE_NONE:` to `default:`

bergzand replied:
See 279e8449cbfb441ceda9f7821ef0ca4a13319a60

dylad replied:
That would be perfect.

dylad commented:
ACK.
arduino-zero is happy. Please squash !

dylad commented:
@bergzand 
Feel free to squash when all static-tests has been fixed.

bergzand commented:
@dylad Squashed, all green here! :tada:

dylad commented:
Here we go !

bergzand commented:
@dylad Thanks for the review and testing!
